✨Performance task locally served pages & wg handlers

micajuine-ho opened the pull request:
~Adding functionality to the performance gulp task to track how long it takes for requests (pageview and ini-load) to fire for pages with amp-analytics tags.~

This PR is now for supporting locally served pages, along with building a structure for wgs easily add in handlers to measure requests (i.e. wg-ads/wg-analytics measuring delay til the `visible` ping is sent out). This is one of the steps in adding the performance task into our CI (since our test pages will be locally served pages).

Changes:
- Config changes to support handlers for different wgs
  - Each handler will test their own defined metrics and include values relevant to those measurements (timeout value, threshold, etc...)
- Changes to `rewrite-script-tags`:
   - Parse script tags of locally hosted websites and download the correct files from CDN or local
   - Download documents that are not explicitly stated in script tags (auto-amp-lightbox-0.1.js)
- Creation of `rewrite-analytics-tags`:
   - Add special identifier to inline config to track visibility pings
   - Fetch and merge vendor configs, that would otherwise fail
- Changes to `measureDocuments`:
   - Capture all outgoing networks requests
   - Apply request handlers, timeout promise, and metric measurer based upon the url being tested
   - Redirect requests for documents that are not explicitly stated in script tags (auto-amp-lightbox-0.1.js), to our downloaded version

zhouyx commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -156,24 +206,33 @@ async function measureDocument(url, version, {headless}) {
   try {
     await page.goto(`file:${urlToCachePath(url, version)}`, {
-      waitUntil: 'networkidle0',
+      waitUntil: 'networkidle2',
```
I read: networkidle2 - consider navigation to be finished when there are no more than 2 network connections for at least 500 ms.

Compared to `networkidle0`, this will wait less time? Is this expected?

micajuine-ho replied:
It is less wait time, but actually I think that `networkidle0` is what we want, because it will give the most time for analytics events to be sent out.

zhouyx commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -118,6 +151,23 @@ const readMetrics = page =>
+function maybeAddAnalyticsMetric(startTime, endTimes, metrics) {
+  if (endTimes.length) {
+    const analyticsRequest =
+      endTimes.reduce((a, b) => a + b, 0) / endTimes.length - startTime;
```
is `endTimes` a shared array? we only calculate the delay for the first request, when would it length be greater than 0.

micajuine-ho replied:
`endTimes` is an array that holds all the times that analytics requests gets fired. There are pages that use more than 1 analytics config, with each config having their own `visible` request, so we should account for these cases.

zhouyx replied:
In that case, it's possible that we are accounting for requests other than `visible` request. For example a timer request. 

IMO, first request would be sufficient to find regression, and less flaky.

micajuine-ho replied:
Ok, I am ok with that since we are averaging out 100 runs (default).

zhouyx commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -70,6 +74,35 @@ const setupMeasurement = page =>
+    });
+  } else if (interceptedUrl.includes(ANALYTICS_PARAM)) {
+    endTimes.push(Date.now());
+    interceptedRequest.abort();
```
Could you help point out where are we waiting for `interceptedRequest` to abort, and what is the timeout?

micajuine-ho replied:
We are aborting because we don't want the request ping to actually hit their endpoint. And the timeout is set by `networkidle0`.

Does that answer your questions?

zhouyx replied:
I see. That makes sense thanks.

zhouyx commented on build-system/tasks/performance/README.md:
```diff
@@ -2,6 +2,6 @@
 Use `gulp performance` to run this test. It measures performance for the current branch compared to the current release. By default, the command first needs to compile minified runtime and components from the current branch by executing the `gulp dist` task. To skip this step, use the `--nobuild` flag, but the task will throw an error if the dist files are missing. Only the runtime and components ...
 
-The `config.json` file contains settings. If you set `headless` to `false`, you will be able to watch the test run in Chrome. The default value for `runs` is 100, which takes ~10 minutes to run, and generates results with ~5% margin of error. Differences larger than this should be investigated. The `urls` array of strings must be set before running the test, and each URL will be tested separately ...
+The `config.json` file contains settings. If you set `headless` to `false`, you will be able to watch the test run in Chrome. The default value for `runs` is 100, which takes ~10 minutes to run, and generates results with ~5% margin of error. Differences larger than this should be investigated. The `urls` array of strings must be set before running the test, and each URL will be tested separately ...
```
I don't think we should ask all to use `gulp serve --compiled` when testing locally.  
It should be up to user to decide to use `gulp serve --compiled` or `gulp serve --cdn`.

micajuine-ho replied:
Well if you use the `--compiled` version the task won't work for locally served pages. It only works with `--cdn`.

estherkim replied:
Why doesn't it worked with `--compiled`?

micajuine-ho replied:
To download the amp scripts ([local](https://github.com/ampproject/amphtml/blob/master/build-system/tasks/performance/rewrite-script-tags.js#L40) and [master](https://github.com/ampproject/amphtml/blob/master/build-system/tasks/performance/rewrite-script-tags.js#L60)), we look for the CDN_URL (`https://cdn.ampproject.org/`), and that's only present when serving with `--cdn`

estherkim replied:
Hmm I see, that's interesting.

zhouyx replied:
I see. I wonder why don't we use the local server proxy `/proxy`.

kevinkimball commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -70,6 +74,35 @@ const setupMeasurement = page =>
+async function handleRequests(interceptedRequest, endTimes) {
+  const interceptedUrl = interceptedRequest.url();
+  const matchArray = interceptedUrl.match(CDN_ANALYTICS_REGEXP);
+  if (matchArray) {
```
Wondering if this is a good opportunity to setup an extensible pattern here, possibly some mapping of URLs to handlers since it seems likely there would be more cases?

zhouyx replied:
That's a good point! I know the ads team want to collect two requests delay. Not sure if they'll use this specific task. Maybe useful to expand the feature here. 
@powerivq Any comment? Thanks

powerivq replied:
I am thinking of this: each team still implements its own handleRequest to respond the payload it desires, and we make another interceptor to record the timing of the request. Does it make sense? @zhouyx

micajuine-ho replied:
@kevinkimball good point!

micajuine-ho replied:
@powerivq I tested out multiple request handlers and it seems to work fine. Seems like a good route to go down in the future.

zhouyx replied:
Looked at the definition of `networkidle0`. I don't know if this would satisfy ads team's request. ActiveView requires the ad to be visible for at least 50% for 1 second. It's very possible that the task ends before activeView. @kevinkimball @micajuine-ho any workaround?

micajuine-ho replied:
I believe just adding a [setTimeout](https://github.com/puppeteer/puppeteer/issues/651) would do the trick, however I'm not sure if we want to always include it. Maybe we could add a flag (`--set-timeout=500`) that would set how long to wait for, if any. Thoughts @zhouyx ?

zhouyx replied:
A race between timeout, and the analytics request sounds good to me. 
My recommendation is to always include the timeout, but set it to a value that's long enough. People will go with the default setting most of the time, and a timeout will serve as a fallback and reduce flakiness.

BTW this is something I'm testing with, please ignore the console log part : )

```
let timeout_count = 0;
const promise = new Promise(resolve => {
    r = resolve;
    setTimeout(() => {
      timeout_count++;
      resolve();
   }, TIMEOUT);
});
  page.on('console', message => {
    if (String(message._text).includes('request delay')) {
      const delay = Number(message._text.replace('analytics request delay is  ', ''));
      all.push(delay);
      r();
    }
  });
.....
await promise; 
```

micajuine-ho replied:
@zhouyx Ok that sounds good to me

zhouyx replied:
Add @estherkim for the best way to customize timeout for different test cases.

rsimha commented:
Adding @estherkim to review changes to the `gulp performance` task.

estherkim commented on build-system/tasks/performance/README.md:
```diff
@@ -2,6 +2,6 @@
-The `config.json` file contains settings. If you set `headless` to `false`, you will be able to watch the test run in Chrome. The default value for `runs` is 100, which takes ~10 minutes to run, and generates results with ~5% margin of error. Differences larger than this should be investigated. The `urls` array of strings must be set before running the test, and each URL will be tested separately ...
+The `config.json` file contains settings. If you set `headless` to `false`, you will be able to watch the test run in Chrome. The default value for `runs` is 100, which takes ~10 minutes to run, and generates results with ~5% margin of error. Differences larger than this should be investigated. The `urls` array of strings must be set before running the test, and each URL will be tested separately ...
 
 This test measures the following metrics: `visible`, `first paint`, `first contentful paint`, `largest contentful paint`, `time to interactive`, `max first input delay`, and `cumulative layout shift`.
```
nit: update this list with analytics metrics

estherkim commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -30,6 +35,8 @@ function requirePuppeteer_() {
   puppeteer = require('puppeteer');
 }
 
+const TIMEOUT_MS = argv.set_timeout || 2000;
```
How about moving the default timeout to config.json?

micajuine-ho replied:
Good call!

zhouyx commented on build-system/tasks/performance/config.json:
```diff
@@ -1,5 +1,6 @@
 {
   "headless": true,
   "runs": 100,
+  "timeout": 2000,
```
If I can ask, this will apply 2000 timeout to all tests with requestHandler. How would you customize it for different use case?

estherkim replied:
Hmm not sure. Can you give an example where another timeout is needed? Do you think it will happen often?

micajuine-ho replied:
Discussed offline, agreed to try out having a config section of each request handler to specify their  metrics, such as URLs and timeout.

zhouyx commented on build-system/tasks/performance/index.js:
```diff
@@ -28,11 +29,12 @@ const {printReport} = require('./print-report');
   await compileScripts(urls);
   await rewriteScriptTags(urls);
-  await getMetrics(urls, {headless, runs});
+  await rewriteAnalyticsTags(urls);
```
I think `rewriteScriptTags` and `rewriteAnalyticsTags` can be merged to one to avoid read and write files multiple times. 
Another optimization would be to not invoke `rewriteAnalyticsTags` for urls without the analytics handler defined. 
Not blocking, but may be a good optimization to have once we start to run the task on travis. cc @estherkim

micajuine-ho replied:
> Another optimization would be to not invoke rewriteAnalyticsTags for urls without the analytics handler defined.

Sounds good to me.

> I think rewriteScriptTags and rewriteAnalyticsTags can be merged to one to avoid read and write files multiple times. 

Maybe once the ads team checks in their handler, we can see how we can group the rewrites (assuming that they need to rewrite something).

zhouyx commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -70,6 +75,51 @@ const setupMeasurement = page =>
+async function handleAnalyticsRequests(interceptedRequest, setEndTimeCallback) {
+  const interceptedUrl = interceptedRequest.url();
+  const matchArray = interceptedUrl.match(CDN_ANALYTICS_REGEXP);
+  if (matchArray) {
```
IMO, this is required for all tests, not only ones with analytics request handler. What do you think?

micajuine-ho replied:
Hmm, I don't have the same initial inclination. Can you explain your reasoning?

zhouyx replied:
My reasoning is that most pages we are testing with will include `<amp-analytics>` component, and loading the analytics vendor config would be a requirement for all test urls.

micajuine-ho replied:
Ok, so then are you thinking that the analyticsHandler is default included with every url?

zhouyx replied:
Nope. 
What if I'm testing with `some-url.html` with no analyticsHandler, but the document includes `<amp-analytics type='something'>`. How would you fake the vendor request in that case?

micajuine-ho replied:
Ooooohhhh I see what you're saying, yea totally agree. Will add it in.

zhouyx commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -148,32 +243,43 @@ function writeMetrics(url, version, metrics) {
 
-  const metrics = await readMetrics(page);
+  let metrics = await readMetrics(page);
+  await delayBasedOnHandlerOptions(handlerOptions);
```
If I read it correctly the process looks like this
wait for all performance metrics to be collected
wait for another timeout value no matter what.

While I'm not sure if there's a timeout for `readMetrics`, but I think we should definitely return early once the handler metric has already been collected without waiting for the timeout.

micajuine-ho replied:
Ok that sounds good to me.

micajuine-ho commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -148,32 +243,43 @@ function writeMetrics(url, version, metrics) {
-  } catch {
+  } catch (e) {
     // site did not load
+    console.log(e);
```
@kevinkimball I found this really helpful when debugging and it could be helpful for other teams as they create the own request handlers. Do you think it's worth keeping?

kevinkimball replied:
Seems OK to me, or could be hidden behind a logging flag

estherkim replied:
If it didn't load, the test will fail right? How about throwing an error after closing the browser?

micajuine-ho replied:
I think a log flag would be good, and then logging the error after the browser closes.

micajuine-ho commented on build-system/tasks/performance/config.json:
```diff
@@ -1,5 +1,15 @@
   "headless": true,
   "runs": 100,
-  "urls": []
+  "urls": [],
```
Yes, that's what it does. Would it be more clear to have `noHandlerUrls: []` instead of making it an object, because right now I don't see any other config data that fits in to the `noHandlers` category (`headless` and `runs` applies to *all* urls)?

zhouyx replied:
I find the declaration a bit confusing because `headless` and `runs` applies to all urls, but the top level `urls` doesn't. The mix and match makes the configuration structure not intuitive imo. 

I don't have strong opinion as long as everything agrees with the config setting.

estherkim replied:
Hmm what do you guys think of the following:
```
urls: {
  'http://normal.test',
  'http://analytics-1.test': {handler: analyticsHandler},
  'http://analytics-2.test': {handler: analyticsHandler},
  'http://ads-1.test': {handler: adsHandler},
}
```

I got the inspo from eslint configs.

micajuine-ho replied:
@estherkim I like it, but would it have `'http://normal.test': {},`?

estherkim replied:
Yeah hmm. Could be ok?

zhouyx replied:
I would still prefer categorize by handler, because it is more often that one wants to test all handler related urls locally. But this would also work!

micajuine-ho replied:
@kevinkimball Any preference between 
```
urls: {
  'http://normal.test',
  'http://analytics-1.test': {handler: analyticsHandler},
  'http://analytics-2.test': {handler: analyticsHandler},
  'http://ads-1.test': {handler: adsHandler},
}
```
vs
```
"noHandlers": {
  "urls": [],
},
"analyticsHandler": {
  "urls": [],
},
```

micajuine-ho commented on build-system/tasks/performance/rewrite-analytics-tags.js:
```diff
@@ -0,0 +1,75 @@
+ * @param {!Array<string>} urls
+ * @return {Promise}
+ */
+function rewriteScriptTags(urls) {
```
oops, I just forgot to change the name. Going with`rewriteAnalyticsConfig`.

micajuine-ho commented:
Ready for another round of review. Changes since last review: 
- Config change to include `defaultHandler`
- Rewrite-script changes to find locally served scripts and correctly pull them from the CDN or dist
- Refactor `measureDocuments` to only have 1 request listener and then call upon sub handlers (for analytics, ads, etc...)
- Added `requestsFailed` metric to give us more insight into the task

zhouyx commented on build-system/tasks/performance/index.js:
```diff
@@ -28,11 +29,12 @@ const {printReport} = require('./print-report');
 async function performance() {
   installPackages(__dirname);
-  const {headless, runs, urls} = new loadConfig();
+  const {headless, runs, urls, handlers} = new loadConfig();
```
This is so so confusing. 
We first flat the config object to an array of urls and the raw handlers config. 
Then later in `measureDocument` we iterate the handler config and find the corresponding handler type again? 
Could you please optimize the design a bit. Thanks.

micajuine-ho commented on build-system/tasks/performance/helpers.js:
```diff
@@ -18,7 +18,16 @@ const fetch = require('node-fetch');
+const EXTRA_URL_PARAM = {
+  'analytics': 'amp-analytics-performance-param',
+};
+const ANALYTICS_PARAM = Object.keys(EXTRA_URL_PARAM)
```
Yes, good call

micajuine-ho commented on build-system/tasks/performance/helpers.js:
```diff
@@ -18,7 +18,16 @@ const fetch = require('node-fetch');
+  .toString();
 const CDN_URL = 'https://cdn.ampproject.org/';
+const AMP_JS_PATH = '/dist/amp.js';
+const LOCAL_PATH_REGEXP = /dist\/(v0\/amp-[A-Za-z\-0-9\.]+).max(.js)/;
```
Originally, we thought that `gulp --cdn` to run local pages would work, since that changes the script tags to `cdn.ampproject....` and we can parse out the filepath. 

However, to use `gulp --cdn`, we have to run `gulp dist --fortesting`, which then causes `amp-analytics/config.js` to not work correctly; it requests `file://...v0/analytics-vendors/gtag.json`, which is blocked because AMP doesn't allow requests using the file protocol (it uses the file protocol because `this.win.location.protocol` is file on puppeteer.

So the fix is to use neither gulp --cdn to serve local pages, nor gulp --fortesting, but just parse out the filepaths manually. Also the change wasn't too bad.

micajuine-ho replied:
> But this doesn't explain why you need .max.js and also amp.js, they should never appear when you run gulp dist.

Turns out I was making mistakes when running the `gulp dist --forstesting` + `gulp serve --compiled`.

micajuine-ho commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -118,6 +179,80 @@ const readMetrics = (page) =>
+ */
+function setUpRequestHandler(handlerOptions, handlerName, page) {
+  // Default handlers
+  const handlers = [analyticsVendorRewriteHandler];
```
Yup, it's default because we should want to keep control and experiment as similar as possible (configs shouldn't change between versions). We can setup the default handler inside measure documents and add an explanation.

zhouyx commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -118,6 +181,95 @@ const readMetrics = (page) =>
+ */
+function setUpAddionalHandlers(handlersList, handlerOptions) {
+  // Setting up timing and adding specified handler
+  if (handlerOptions.handlerName === 'analyticsHandler') {
```
Correct me if this wrong. 
All urls within the same `handlers` config share the same `handlerOptions` config. (we're not creating new instances of the `handlerOptions` object within `load-config.js` file). I'm afraid rewriting the `handlerOptions` object would affect other instances that fall within the same handler. Have you tried the test with multiple urls.

micajuine-ho replied:
Yes I have tried it with multiple urls; within `measureDocument()`, we create `handlerOptionsForUrl` by creating a shallow copy of the original `handlerOptions` config. This works fine for storing the `startTime` and `endTime`, but would cause some bugs if someone were to mutate objects within the `handlerOptions`. Instead of shallow copy, should I create a deep copy?

zhouyx commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -147,33 +299,43 @@ function writeMetrics(url, version, metrics) {
   }
 
   const metrics = await readMetrics(page);
+  await delayBasedOnHandlerOptions(handlerOptionsForUrl);
```
If a request arrives right after you start the wait here, you would still need to wait for the entire timeout period. Please make this a promise race between request arrives vs timeout  with a defined startpoint.

micajuine-ho replied:
Will do

zhouyx commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -147,33 +299,43 @@ function writeMetrics(url, version, metrics) {
+  await page.setRequestInterception(true);
   await setupMeasurement(page);
+  setUpDefaultHandlers(handlersList);
+  setUpAddionalHandlers(handlersList, handlerOptionsForUrl);
```
nit: additionalHandlers

micajuine-ho commented on build-system/tasks/performance/load-config.js:
```diff
@@ -22,15 +22,25 @@ const PATH = './config.json';
+    });
+    return mapping;
+  }, {});
+  if (Object.keys(config.urlToHandlers).length < 1) {
```
Yes and yes. Plan is that before we integrate with our CI, we add urls.

micajuine-ho commented on build-system/tasks/performance/load-config.js:
```diff
@@ -22,15 +22,25 @@ const PATH = './config.json';
+      handlerOptions.urls = [argv.url];
+    }
+    handlerOptions.urls.forEach((url) => {
+      if (mapping[url]) {
```
Yup, that sounds good to me!

micajuine-ho commented on build-system/tasks/performance/config.json:
```diff
@@ -1,5 +1,24 @@
+      "extraUrlParam": {
+        "analytics": "amp-analytics-performance-param"
+      },
+      "analyticsParam": "analytics=amp-analytics-performance-param"
```
@zhouyx Agree with removing `analyticsParam` from the config.

> extraUrlParam and analyticsParam don't apply to all handlers.

Yep, they already live in the `analyticsHandler`. Is that good enough?

zhouyx commented on build-system/tasks/performance/README.md:
```diff
@@ -2,6 +2,6 @@
+The `config.json` file contains settings. If you set `headless` to `false`, you will be able to watch the test run in Chrome. The default value for `runs` is 100, which takes ~10 minutes to run, and generates results with ~5% margin of error. Differences larger than this should be investigated. The `defaultHandler` within the `handlers` list measures the basic pageview metrics (visible, first pai ...
 
-This test measures the following metrics: `visible`, `first paint`, `first contentful paint`, `largest contentful paint`, `time to interactive`, `max first input delay`, and `cumulative layout shift`.
+This test measures the following metrics: `visible`, `first paint`, `first contentful paint`, `largest contentful paint`, `time to interactive`, `max first input delay`, `cumulative layout shift`, and `analytics request time`.
```
nit: request delay?

zhouyx commented on build-system/tasks/performance/README.md:
```diff
@@ -2,6 +2,6 @@
 Use `gulp performance` to run this test. It measures performance for the current branch compared to the current release. By default, the command first needs to compile minified runtime and components from the current branch by executing the `gulp dist` task. To skip this step, use the `--nobuild` flag, but the task will throw an error if the dist files are missing. Only the runtime and components ...
 
-The `config.json` file contains settings. If you set `headless` to `false`, you will be able to watch the test run in Chrome. The default value for `runs` is 100, which takes ~10 minutes to run, and generates results with ~5% margin of error. Differences larger than this should be investigated. The `urls` array of strings must be set before running the test, and each URL will be tested separately ...
+The `config.json` file contains settings. If you set `headless` to `false`, you will be able to watch the test run in Chrome. The default value for `runs` is 100, which takes ~10 minutes to run, and generates results with ~5% margin of error. Differences larger than this should be investigated. The `defaultHandler` within the `handlers` list measures the basic pageview metrics (visible, first pai ...
```
nit: analytics request delay, request timeout count

zhouyx commented on build-system/tasks/performance/helpers.js:
```diff
@@ -19,13 +19,16 @@ const fs = require('fs');
 const CDN_URL = 'https://cdn.ampproject.org/';
+const V0_PATH = '/dist/v0.js';
+const LOCAL_PATH_REGEXP = /dist\/(v0\/amp-[A-Za-z\-0-9\.]+.js)/;
+const ANALYTICS_VENDORS_PATH = '../../../extensions/amp-analytics/0.1/vendors/';
```
Let's use the compiled vendor which lives under v0/analytics-vendors/xxx.json

micajuine-ho replied:
Ok, if we're using that path then we should also add the `gulp vendor-configs` command as part of this task...

That command will be run in the Travis build, but I think it's good to have it for local testing.

zhouyx replied:
I believe we always run that with `gulp build` or `gulp dist`, so it should be fine?

zhouyx commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -118,6 +196,124 @@ const readMetrics = (page) =>
+    case 'analyticsHandler':
+      // Set up timing
+      Object.assign(handlerOptions, {'startTime': Date.now()});
+      handlersList.push((interceptedRequest) =>
```
are we returning a promise here with analyticsHandler?

micajuine-ho replied:
Adding the `analyticsHandler` to the `handlerList`. The `analyticsHandler` returns a promise to signal if it actually did anything with the request.

zhouyx commented on build-system/tasks/performance/measure-documents.js:
```diff
@@ -118,6 +196,124 @@ const readMetrics = (page) =>
+            if (!handlerOptions.endTime) {
+              Object.assign(handlerOptions, {
+                endTime,
+                'timeout': 0,
```
To confirm you've tried with multiple urls, and they work fine? Asking again because my understanding is this would override the handlerOptions['timeout'] to 0, and would affect the next test. 
There's could be a deep copy somewhere that I missed.

micajuine-ho replied:
Yup tried it multiple urls. Each URL has a copy of the handlerOptions so no worry about mistakenly overriding something.

zhouyx replied:
Thanks for confirming

zhouyx commented on build-system/tasks/performance/rewrite-analytics-tags.js:
```diff
@@ -0,0 +1,112 @@
+    tag.removeAttribute('type');
+    const vendorConfig = await getVendorConfig(vendor);
+    // Shallow copy is ok
+    const x = Object.assign(vendorConfig, script);
```
Let's keep this for now, but you'll need a deep merge. Is it possible to import method from amp-analytics/config.js?

micajuine-ho replied:
I think generally we don't want to rely on v0 for stuff within our build-system? @estherkim can you confirm?

estherkim replied:
It's okay to import the method. We import from extensions when compiling parsers, for one. I also found some instances that import `src/mode`, `src/layout` and more.

micajuine-ho replied:
Ok, thanks @estherkim. I'll just add it in now.

zhouyx replied:
I don't see the change. Could you add a TODO?

micajuine-ho commented:
@estherkim  PTAL

zhouyx commented on build-system/tasks/performance/compile-scripts.js:
```diff
@@ -28,6 +29,7 @@ async function compileScripts(urls) {
   if (!argv.nobuild) {
```
--nobuild is only used when you have already built the binary and don't want to build it again. 
There's no need to await for `vendorConfigs`.
